tests/test_dockerspawner.py

```diff
@@ -82,27 +82,48 @@ async def test_start_stop(dockerspawner_configured_app, remove):
         assert sorted(state) == ["object_id", "object_name"]
 
 
-@pytest.mark.parametrize("image", ["1.0", "1.1.0", "nomatch"])
-async def test_allowed_image(dockerspawner_configured_app, image):
+def allowed_images_callable(*_):
+    return ["jupyterhub/singleuser:1.0", "jupyterhub/singleuser:1.1"]
+
+
+@pytest.mark.parametrize(
+    "image, expected",
+    [
+        (
+            {
+                "1.0": "jupyterhub/singleuser:1.0",
+                "1.1": "jupyterhub/singleuser:1.1",
+            },
+            "1.0",
+        ),
+        (["jupyterhub/singleuser:1.0", "jupyterhub/singleuser:1.1.0"], "1.1.0"),
+        (allowed_images_callable, "1.0"),
+    ],
+)
+async def test_allowed_image(dockerspawner_configured_app, image, expected):
     app = dockerspawner_configured_app
     name = "checker"
     add_user(app.db, app, name=name)
     user = app.users[name]
     assert isinstance(user.spawner, DockerSpawner)
     user.spawner.remove_containers = True
-    user.spawner.allowed_images = {
-        "1.0": "jupyterhub/singleuser:1.0",
-        "1.1": "jupyterhub/singleuser:1.1",
-    }
+    user.spawner.allowed_images = image
     token = user.new_api_token()
     # start the server
     r = await api_request(
-        app, "users", name, "server", method="post", data=json.dumps({"image": image})
+        app,
+        "users",
+        name,
+        "server",
+        method="post",
+        data=json.dumps({"image": expected}),
     )
-    if image not in user.spawner.allowed_images:
-        with pytest.raises(Exception):
-            r.raise_for_status()
-        return
+
+    if not callable(image):
+        if expected not in user.spawner.allowed_images:
```

Comment:
Instead of skipping this check, let's compute the allowed_images dict if it's callable:

```suggestion
    if image not in user.spawner._get_allowed_images():
```

(`_get_allowed_images()` is always correct whether `allowed_images` is callable or not)


This skip is why the test is failing - if `image` is not in `allowed_images`, the request fails and the test stops here (which is the case for your callable test - it would be good to have both an allowed and rejected case for the callable). But since the short-circuit is skipped if callable config is used, the right branch is not taken.

Reply:
Great insight! Bug is fixed!